R/ptype.R

```diff
@@ -111,3 +111,90 @@ recipes_ptype <- function(x, ..., stage = "prep") {
   
   ptype
 }
+
+#' Validate prototype of recipe object
+#'
+#' This helper function validates a dataframe against the ptype of a recipe.
+#'
+#' @param x A `recipe` object.
+#' @param new_data A data.frame. To be patched aganist ptype of `x`.
+#' @param ... currently not used.
+#' @param stage A single character. Must be one of `"prep"` or `"bake"`. See
+#'   details for more. Defaults to `"prep"`.
+#' @param call The execution environment of a currently running function, e.g.
+#'   `caller_env()`. The function will be mentioned in error messages as the
+#'   source of the error. See the call argument of [rlang::abort()] for more
+#'   information.
+#' 
+#' @return Nothing or an error.
+#' @keywords internal
+#'
+#' @seealso [developer_functions] [recipes_ptype]
+#' 
+#' @examples
+#' rec <- recipe(mpg ~ disp, data = mtcars)
+#' 
+#' recipes_ptype_validate(rec, mtcars)
+#' 
+#' @export
+recipes_ptype_validate <- function(x,
+                                   new_data,
+                                   ...,
+                                   stage = "prep", 
+                                   call = rlang::caller_env()) {
+  old_ptype <- recipes_ptype(x, stage = stage)
+  col_names <- names(old_ptype)
+
+  new_ptype <- vctrs::vec_ptype(new_data)
+
+  if (!all(col_names %in% names(new_ptype))) {
+    offenders <- col_names[!col_names %in% names(new_ptype)]
+    cli::cli_abort(
+      "Not all variables in the recipe are present in the supplied training \\
+      set: {.var {offenders}}.",
+      call = call
+    )
+  }
+
+  new_ptype <- new_ptype[col_names]
+
+  if (!identical(lapply(old_ptype, class), lapply(new_ptype, class))) {
+    old_classes <- lapply(old_ptype, class)
+    new_classes <- lapply(new_ptype, class)
+
+    offenders <- purrr::map2_lgl(old_classes, new_classes, identical)
+    offenders <- col_names[!offenders]
+
+    msg <- c("x" = "{cli::qty(offenders)} The following variable{?s} \\
+                   has the wrong class:")
+
+    col_msg <- paste0(
+      "{.var {offenders[",
+      seq_along(offenders),
+      "]}} must have class {.cls {new_classes[offenders[",
+      seq_along(offenders),
+      "]]}} not {.cls {old_classes[offenders[",
+      seq_along(offenders),
+      "]]}}."
+    )
+    names(col_msg) <- rep("*", length(col_msg))
+    msg <- c(msg, col_msg)                                                           
+
+    cli::cli_abort(msg, call = call)
+  }
+    
+  if (!identical(lapply(old_ptype, attributes), lapply(new_ptype, attributes))) {
+    old_attributes <- lapply(old_ptype, attributes)
+    new_attributes <- lapply(new_ptype, attributes)
+  
+    offenders <- purrr::map2_lgl(old_attributes, new_attributes, identical)
+    offenders <- col_names[!offenders]
+  
+    msg <- c("x" = "{cli::qty(offenders)} The following variable{?s} \\
+                   has the wrong attributed: {.var {offenders}}.")
```

Comment:
```suggestion
                   {?has/have} the wrong attributes: {.var {offenders}}.")
```

Again, not sure if I have the right pluralization order here.

Reply:
In what cases do these attributes matter? Do we need to enforce that they not change?

Especially given that the message doesn't diagnose which attributes differ, I think I'd be pretty lost and frustrated if I saw this message.

Reply:
Examples would be time zone in datetime variables and factor levels are both stored as attributes. Which i think should both be checked

``` r
as.POSIXct(0, origin = "1960-01-01") |> attributes()
#> $tzone
#> [1] ""
#> 
#> $class
#> [1] "POSIXct" "POSIXt"
as.POSIXct(0, origin = "1960-01-01", tz = "GMT") |> attributes()
#> $tzone
#> [1] "GMT"
#> 
#> $class
#> [1] "POSIXct" "POSIXt"

factor(1:5) |> attributes()
#> $levels
#> [1] "1" "2" "3" "4" "5"
#> 
#> $class
#> [1] "factor"
factor(1:4) |> attributes()
#> $levels
#> [1] "1" "2" "3" "4"
#> 
#> $class
#> [1] "factor"
```

Reply:
i do agree that the error isn't great, the problem i ran into was it was hard to explain the difference as attributes could be anything

Reply:
what about?
- special case errors for factors and datetimes (most common)
- then more generic error, with code snippet as to how to figure out what went wrong?

Reply:
Sure, that sounds good!